alchemiscale/storage/statestore.py

```diff
@@ -242,138 +242,139 @@ def reset(self):
-                        labels=["GufeTokenizable", value.__class__.__name__],
-                        gufe_key=value.key,
-                        scope=scope,
-                    )
-                    self.gufe_nodes[
-                        (value.key, scope.org, scope.campaign, scope.project)
-                    ] = node_
-                subgraph = (
-                    subgraph
-                    | Relationship.type("DEPENDS_ON")(
-                        node,
-                        node_,
-                        attribute=key,
-                        _org=scope.org,
-                        _campaign=scope.campaign,
-                        _project=scope.project,
+            ) is None:
+                raise ValueError("Possibly corrupt keyedchain")
+            return rel_node
+
+        def update_relationships(node_a, node_b, **kwargs):
+            rel = Relationship.type("DEPENDS_ON")(
+                node_a,
+                node_b,
+                _org=scope.org,
+                _campaign=scope.campaign,
+                _project=scope.project,
+                **kwargs,
+            )
+            relationships.append(rel)
+
+        def handle_list(node, key, values):
+            if isinstance(values[0], (int, float, str)) and all(
+                (isinstance(x, type(values[0])) for x in values)
+            ):
+                node[key] = values
+                return
+            # list of gufe key: [{":gufe-key:": ...}, {":gufe-key:": ...}, {":gufe-key:": ...}]
+            elif all(map(is_gufe_dict, values)):
+                for i, x in enumerate(values):
+                    rel_key = x[":gufe-key:"]
+                    update_relationships(
+                        node, get_previous_node(rel_key), attribute=key, index=i
                     )
-                    | subgraph_
-                )
+            else:
+                node[key] = json.dumps(values, cls=JSON_HANDLER.encoder)
+                node["_json_props"].append(key)
 
+        def handle_dict(node, key, dct):
+            # e.g. {'ligand': {':gufe-key:': 'SmallMoleculeComponent-abc123'}}
+            if all(map(is_gufe_dict, dct.values())):
+                for k, v in dct.items():
+                    rel_key = v[":gufe-key:"]
+                    update_relationships(
+                        node, get_previous_node(rel_key), attribute=key, key=k
+                    )
             else:
-                node[key] = value
+                node[key] = json.dumps(dct, cls=JSON_HANDLER.encoder)
+                node["_json_props"].append(key)
 
-        subgraph = subgraph | node
+        def handle_tuple(node, key, values):
+            if not (
+                isinstance(values[0], (int, float, str))
+                and all((isinstance(x, type(values[0])) for x in values))
+            ):
+                node[key] = json.dumps(values, cls=JSON_HANDLER.encoder)
+                node["_json_props"].append(key)
```

Comment:
I realize now that this captures the previous behavior, but it looks like we weren't handling the case where this `if` statement gives `False`. In that case, the `tuple` will be lost entirely.

Do you think this is a legit bug @ianmkenney?

Reply:
We should probably just JSONify all `tuple`s to distinguish them from the way `list`s are handled in `_node_to_gufe`.

Reply:
Definitely a bug in both implementations. I forgot to drop a todo in there to get back to it. I'll look into a solution for it tomorrow.